src/main/java/org/broadinstitute/hellbender/tools/walkers/realignmentfilter/Realigner.java

```diff
@@ -0,0 +1,72 @@
+package org.broadinstitute.hellbender.tools.walkers.realignmentfilter;
+
+
+import org.broadinstitute.hellbender.utils.bwa.BwaMemAligner;
+import org.broadinstitute.hellbender.utils.bwa.BwaMemAlignment;
+import org.broadinstitute.hellbender.utils.bwa.BwaMemIndex;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Realigner {
+    private final BwaMemAligner aligner;
+    private final int maxReasonableFragmentLength;
+    private final int minAlignerScoreDifference;
+    private final int numberOfRegularContigs;
+
+    public Realigner(final RealignmentArgumentCollection rfac) {
+        final BwaMemIndex index = new BwaMemIndex(rfac.bwaMemIndexImage);
+        maxReasonableFragmentLength = rfac.maxReasonableFragmentLength;
+        minAlignerScoreDifference = rfac.minAlignerScoreDifference;
+        numberOfRegularContigs = rfac.numRegularContigs;
+        aligner = new BwaMemAligner(index);
+        aligner.setMinSeedLengthOption(rfac.minSeedLength);
+        aligner.setDropRatioOption((float) rfac.dropRatio);
+        aligner.setSplitFactorOption((float) rfac.splitFactor);
+        aligner.setFlagOption(BwaMemAligner.MEM_F_ALL);
+    }
+
+    public RealignmentResult realign(final GATKRead read) {
+        final List<BwaMemAlignment> alignments = aligner.alignSeqs(Arrays.asList(read), GATKRead::getBasesNoCopy).get(0);
+
+        final List<BwaMemAlignment> nonAltAlignments = alignments.size() == 1 ? alignments :
+                alignments.stream().filter(a -> a.getRefId() < numberOfRegularContigs).collect(Collectors.toList());
+        return checkAlignments(nonAltAlignments);
+    }
+
+    public RealignmentResult checkAlignments(final List<BwaMemAlignment> alignments) {
+        if (alignments.isEmpty()) {
+            return new RealignmentResult(false, Collections.emptyList());
+        } else if (alignments.get(0).getRefId() < 0) {
+            return new RealignmentResult(false, alignments);
+        } else if (alignments.size() == 1) {
+            return new RealignmentResult(true, alignments);
```

Comment:
If there is only one alignment, is it really always in the supposed location?

Reply:
Empirically, yes.  I actually wrote a version of this tool that used a liftover chain file to map the hg38 alignment back to the original reference in order to check this, but it slowed the tool, complicated the code, made an additional resource file (which, unlike the hg38 index, depends on the original reference) necessary, and made no difference in the results.

Basically, hg19 is good enough that even if it misses some alignments its best alignment will always show up somewhere in the hg38 alignments.

Reply:
This might not be the case for other reference genomes (e.g., non-model species). I thought that the filter can be used with other species too, which have much worse assemblies...